# LMMEL-miR-miner
MATLAB scripts that use miR target prediction databases (TargetScan 7.1, DIANA-microT CDS) to enrich a statistical analysis 
(Pearson's correlation, mutual information) of miR and mRNA data from the Ludwig Melbourne melanoma (LM-MEL) cell line panel data


This script accompanies the manuscript:
### MC Andrews/J Cursons, DG Hurley, M Anaka, JS Cebon, A Behren, EJ Crampin (2016). Systems analysis identifies miR-29b 
regulation of invasiveness in melanoma. *BMC Molecular Cancer*, (Accepted Nov 2016).
* doi: to-be-assigned


## Accompanying Data

These scripts use a number of data sources, including:

#### LM-MEL Cell Line Panel: miR and mRNA abundance data
* [Original Manuscript]( http://dx.doi.org/10.1111/pcmr.12097 )
* [mRNA abundance data (ArrayExpress)]( https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-1496/ )
* [miR abundance data (GEO)]( http://to.be.fixed )

#### TCGA Skin and Cutaneous Melanoma (SKCM) data: miR and mRNA abundance data from clinical samples
* [Project Link]( http://cancergenome.nih.gov/ )
* [Data Link]( https://tcga-data.nci.nih.gov/tcga/tcgaCancerDetails.jsp?diseaseType=SKCM&diseaseName=Skin%20Cutaneous%20Melanoma )
* [Related Manuscript (Analytical)]( http://dx.doi.org/10.1111/pcmr.12374 )

#### TargetScan v7.1
* [Project Website]( http://http://www.targetscan.org/vert_71/ )
* [Data Link]( http://www.targetscan.org/cgi-bin/targetscan/data_download.vert71.cgi )
* [Agarwal et al., eLife, (2015).]( http://dx.doi.org/10.7554/eLife.05005 )
* [Giang et al., Genes Dev., (2010).]( http://dx.doi.org/10.1146/annurev-genet-120213-092023 )
* [Fromm et al., Annu. Rev. Genet., (2015).]( http://dx.doi.org/10.1146/annurev-genet-120213-092023 )

#### miRTarBase
* [Project Website]( http://http://mirtarbase.mbc.nctu.edu.tw/ )
* [Data Link]( http://mirtarbase.mbc.nctu.edu.tw/php/download.php )
* [Manuscript (2016)]( http://dx.doi.org/10.1093/nar/gkv1258 )
* [Manuscript (2014)]( http://dx.doi.org/10.1093/nar/gkt1266 )

#### DIANA-microT CDS (v5.0)
* [Original Manuscript]
* [Project Website]( http://to-be-fixed )
* [Data Link]( http://to-be-fixed )


### Virtual Reference Environment

For users unfamiliar with python, a Virtual Reference Environment is available for this  
project, containing all scripts, data and documentation in an easily-deployed format.

For further information on Virtual Reference Environments, please refer to the
[Online Documentation](http://uomsystemsbiology.github.io/research/reference-environments/).



## Contacts

For further information, please contact:

#### Dr. Miles Andrews
* 
* *ex*: Cancer Immunobiology Laboratory, Olivia Newton-John Cancer Research Institute, Australia  
* miles.andrews (at) onjcri.org.au

#### Dr. Joe Cursons
* Bioinformatics Division, Walter and Eliza Hall Institute of Medical Research, Australia
* *ex*: Systems Biology Laboratory, University of Melbourne, Australia  
* cursons.j (at) wehi.edu.au

#### Dr. Daniel Hurley 
* 
* *ex*: Systems Biology Laboratory, University of Melbourne, Australia  
* daniel.hurley (at) unimelb.edu.au

#### Dr. Andreas Behren
* Cancer Immunobiology Laboratory, Olivia Newton-John Cancer Research Institute, Australia  
* andreas.behren (at) onjcri.org.au

#### Prof. Jonathan Cebon
* Cancer Immunobiology Laboratory, Olivia Newton-John Cancer Research Institute, Australia  
* jonathan.cebon (at) onjcri.org.au

#### Prof. Edmund Crampin
* Systems Biology Laboratory, University of Melbourne, Australia  
* edmund.crampin (at) unimelb.edu.au
